commons/src/main/java/io/aiven/kafka/tiered/storage/commons/chunkindex/serde/ChunkSizesBinaryCodec.java

```diff
@@ -0,0 +1,199 @@
+ * the last value doesn't participate in the compression and is encoded independently in full 4 bytes.
+ *
+ * <p>The first technique used to reduce the byte count per value is finding the max number of bytes
+ * needed to encode values and using it instead of full 4 bytes.
+ *
+ * <p>Based on the similarity assumption and the first technique, the second technique
+ * is to find the "base" (i.e. the minimum) and subtract it from all values (again, except for the last),
+ * thus reducing the number of bytes needed to encode values.
+ *
+ * <p>For example, consider values <code>1000000, 1000010, 1000020</code> (each needs min 3 bytes to encode).
+ * The base is 1000000 and "de-based" values are <code>0, 10, 20</code>, which need 1 byte per value.
+ *
+ * <p>The average byte count per value depends mostly on the variability of values. For example
+ *
+ * <pre>
+ *     int size = Integer.MAX_VALUE;  // 2 GB
+ *     int originalChunkSize = 1024 * 1024;  // 1 MB
+ *     int variability = 300;
+ *     List&lt;Integer&gt; chunks = new ArrayList<>();
+ *     var random = new Random();
+ *     for (int pos = originalChunkSize; pos < size - originalChunkSize; pos += originalChunkSize) {
+ *         chunks.add(originalChunkSize + random.nextInt(variability));
+ *     }
+ *     System.out.println("Bytes per value: "
+ *         + (float) ChunkSizesBinaryCodec.encode(chunks).length / chunks.size());
+ * </pre>
+ * prints "<code>Bytes per value: 2.0053763</code>".
+ * If variability is <code>200</code>, it prints "<code>Bytes per value: 1.0058651</code>".
+ *
+ * <p>The byte layout is the following:
+ * <table>
+ *     <tr>
+ *         <th>Field</th>
+ *         <th>Size, B</th>
+ *     </tr>
+ *     <tr>
+ *         <td>Count</td>
+ *         <td>4</td>
+ *     </tr>
+ *     <tr>
+ *         <td>Base</td>
+ *         <td>4</td>
+ *     </tr>
+ *     <tr>
+ *         <td>Bytes per value</td>
+ *         <td>1</td>
+ *     </tr>
+ *     <tr>
+ *         <td>Values (without last)</td>
+ *         <td>(Count - 1) * Bytes per value</td>
+ *     </tr>
+ *     <tr>
+ *         <td>Last value</td>
+ *         <td>4</td>
+ *     </tr>
+ * </table>
+ *
+ * <p>When there are zero values, only the count is output.
+ *
+ * <p>When there is one value, only the count and the (last) value is output,
+ * i.e. no base, bytes per value, non-last values.
+ *
+ * <p>The codec accepts only non-negative (>= 0) values.
+ */
+class ChunkSizesBinaryCodec {
+    private static final int COUNT_SIZE = 4;
+    private static final int BASE_SIZE = 4;
+    private static final int BYTES_PER_VALUE_SIZE = 1;
+    private static final int FULL_VALUE_SIZE = 4;
+
+    static byte[] encode(final List<Integer> values) {
+        final int count = values.size();
+        if (count == 0) {
+            final ByteBuffer intBuf = ByteBuffer.allocate(COUNT_SIZE);
+            intBuf.putInt(count);
+            return intBuf.array();
```

Comment:
Should we just use a constant here instead?

Reply:
Nah, I wouldn't bother, it's most likely an impossible event.